libraries/commerce/cart/actions/addProductsToCart.js

```diff
@@ -6,43 +6,58 @@ import addProductsToCart from '../action-creators/addProductsToCart';
 import successAddProductsToCart from '../action-creators/successAddProductsToCart';
 import errorAddProductsToCart from '../action-creators/errorAddProductsToCart';
 import setCartProductPendingCount from '../action-creators/setCartProductPendingCount';
-import { getProductPendingCount } from '../selectors';
+import { getProductPendingCount, getAddToCartMetadata } from '../selectors';
+import { getProductById } from '../../product/selectors/product';
 import { messagesHaveErrors } from '../helpers';
 
 /**
- * Adds a product to the cart.
+ * Adds a product to the cart. Includes metadata if available and not provided before.
  * @param {Array} productData The options for the products to be added.
  * @return {Function} A redux thunk.
  */
 const addToCart = productData => (dispatch, getState) => {
-  const pendingProductCount = getProductPendingCount(getState());
+  const state = getState();
+  const pendingProductCount = getProductPendingCount(state);
+  const products = productData.map(product => {
+    const { productData } = getProductById(state, product.productId);
+    const productId = productData.baseProductId || product.productId;
 
-  dispatch(addProductsToCart(productData));
+    // it's a variant if it has a baseProductId
+    const variantProductId = productData.baseProductId ? productData.id : undefined;
+    const metadata = getAddToCartMetadata(state, productId, variantProductId);
+    return product.metadata && product || {
+      ...product,
+      ...(metadata) && { metadata },
+    }
+  })
+
+  dispatch(addProductsToCart(products));
   dispatch(setCartProductPendingCount(pendingProductCount + 1));
 
   const request = new PipelineRequest(pipelines.SHOPGATE_CART_ADD_PRODUCTS);
-  request.setInput({ products: productData })
+
+  request.setInput({ products })
     .setResponseProcessed(PROCESS_SEQUENTIAL)
     .setRetries(0)
     .dispatch()
     .then(({ messages }) => {
       const requestsPending = request.hasPendingRequests();
 
-      if (messagesHaveErrors(messages)) {
+      if (messages && messagesHaveErrors(messages)) {
```

Comment:
Why is the extra check necessary? The method defaults messages to an empty array.

Reply:
I had cases, where the magento pipeline responses were
```
{
  "messages": null
}
```

This caused errors in the frontend.